File: src/engine/adapters/mysql-family/query.ts
import type { Connection, FieldInfo, TypeCastFunction, TypeCastNextFunction } from 'mariadb';
import { withAbortRace } from '../abort';
import type { OpCtx } from '../adapter';
import { assertNotCancelled } from '../errors';
import { mapError } from './errors';

export interface RunningQuery {
  threadId: number | null;
}

// P13 D3: the tracker registers a running query and hands back its own release, called once the
// statement settles (resolve, reject, or abort) — the only way `runningByOp` shrinks other than
// `cancel()`/`disconnect()`. The release closure itself does the identity check against what is
// currently registered for the op, so a later statement in the same multi-statement op (mutate's
// transaction, console's "Run all") is never unregistered by an earlier one settling after it.
export type TrackQuery = (q: RunningQuery) => () => void;

// The wire protocol has no distinct TEXT type code — TEXT/TINYTEXT/MEDIUMTEXT/LONGTEXT share the
// exact same BLOB-family type codes as their binary counterparts, distinguished only by the
// column's collation (binary collation = a real BLOB; any other = TEXT). GEOMETRY and BIT have no
// such ambiguity — always binary.
const BLOB_FAMILY_TYPES = new Set(['TINY_BLOB', 'MEDIUM_BLOB', 'LONG_BLOB', 'BLOB']);
const ALWAYS_BINARY_TYPES = new Set(['GEOMETRY', 'BIT']);

// column.string() on a binary column decodes as UTF-8 and mangles the bytes — the MariaDB
// equivalent of Postgres's bytea handling (D3). VAR_STRING/STRING/the BLOB family only count as
// binary when the column's own collation is the binary collation (a BINARY/VARBINARY/BLOB column,
// not a TEXT/CHAR/VARCHAR one) — the driver's Collation table names it 'BINARY' (uppercase), so
// this compares case-insensitively rather than assuming a casing.
// Exported so mysql-family/console.ts (P5.5) can reuse the exact same binary/text decoding — the
// query console needs the identical hex-normalisation discipline the read path already has.
export const typeCastString: TypeCastFunction = (field: FieldInfo, _next: TypeCastNextFunction) => {
  const isBinaryString =
    (field.type === 'VAR_STRING' || field.type === 'STRING' || BLOB_FAMILY_TYPES.has(field.type)) &&
    field.collation?.name?.toUpperCase() === 'BINARY';
  if (ALWAYS_BINARY_TYPES.has(field.type) || isBinaryString) {
    const buf = field.buffer();
    return buf === null ? null : `0x${buf.toString('hex')}`;
  }
  return field.string();
};

export interface QueryOptions {
  rowsAsArray?: boolean;
  /** Identity string decoding + binary-to-hex (D3) — the read path only, never catalog queries. */
  textMode?: boolean;
  /** Appends the bound parameter values to the logged command (§5b step 6), read path only. */
  logParams?: boolean;
}

// The cancellable query helper (§5c), mirroring postgres/query.ts exactly. Exported so
// tests/db/mariadb.spec.ts can drive a `SELECT SLEEP(30)` directly to assert server-side
// cancellation without inventing a fake tree level.
export async function runQuery<R = unknown>(
  conn: Connection,
  sql: string,
  params: unknown[],
  ctx: OpCtx,
  track: TrackQuery,
  opts?: QueryOptions,
): Promise<R[]> {
  ctx.setCommand(
    opts?.logParams && params.length > 0 ? `${sql} -- params: ${JSON.stringify(params)}` : sql,
  );

  assertNotCancelled(ctx);

  const release = track({ threadId: conn.threadId });

  const queryOptions: { sql: string; rowsAsArray?: boolean; typeCast?: typeof typeCastString } = {
    sql,
    rowsAsArray: opts?.rowsAsArray,
  };
  if (opts?.textMode) queryOptions.typeCast = typeCastString;
  // Always the text protocol (conn.query), never conn.execute()'s binary/prepared protocol:
  // params still bind safely via query()'s own client-side `?` escaping, and the binary protocol
  // combined with the textMode typeCast callback above corrupts row data (confirmed with real
  // bound params, not just param-less queries).
  const issue = conn.query.bind(conn);

  // The connector has no per-query abort. The AbortSignal listener below only stops *us*
  // waiting — the server-side kill is entirely adapter.cancel()'s job (KILL QUERY over a side
  // connection, using the threadId tracked above); engine/scheduler/ops.ts's cancelOp()
  // triggers both, in that order — the same discipline as Postgres's pg_cancel_backend path.
  const rows = await withAbortRace(ctx, () => issue(queryOptions, params), { release, mapError });
  return rows as R[];
}

export interface CommandOptions {
  /** setCommand() was already called once for the whole batch (P5 D9) — do not call it again. */
  suppressCommand?: boolean;
}

// mutate.ts's counterpart to runQuery: for a non-SELECT statement, `conn.query()` resolves an
// OkPacket-shaped object (`{ affectedRows, insertId, warningStatus }`), not an array of rows —
// this is what every mutate.ts caller actually needs. Cancellation and error-mapping mirror
// runQuery's; only the settled value and the setCommand discipline differ.
export async function runCommand(
  conn: Connection,
  sql: string,
  params: unknown[],
  ctx: OpCtx,
  track: TrackQuery,
  opts?: CommandOptions,
): Promise<{ affectedRows: number }> {
  if (!opts?.suppressCommand) ctx.setCommand(sql);

  assertNotCancelled(ctx);

  const release = track({ threadId: conn.threadId });

  const result = await withAbortRace<{ affectedRows?: number }>(
    ctx,
    () => conn.query({ sql }, params),
    { release, mapError },
  );
  return { affectedRows: result.affectedRows ?? 0 };
}

File: apps/kira-studio/internal/adapters/mysqlfamily/query.go
package mysqlfamily

import (
	"context"
	"database/sql"
	"encoding/json"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
)

// RunningQuery is query.ts's RunningQuery.
type RunningQuery struct {
	ThreadID uint32
}

// TrackQuery is query.ts's TrackQuery: registers a running query and returns its own release,
// called once the statement settles. The identity-checked release lives in the caller
// (adapter.go's trackerFor), same as postgres's.
type TrackQuery func(RunningQuery) (release func())

// QueryOptions is query.ts's QueryOptions, minus rowsAsArray: runArrayQuery below always scans
// positionally, so there is no separate rows-as-object mode to select.
type QueryOptions struct {
	// LogParams appends the bound parameter values to the logged command, read path only.
	LogParams bool
}

func setCommand(op *adapters.OpCtx, sql string, params []any, logParams bool) {
	if logParams && len(params) > 0 {
		if b, err := json.Marshal(params); err == nil {
			op.SetCommand(sql + " -- params: " + string(b))
			return
		}
	}
	op.SetCommand(sql)
}

// runArrayQuery is read.ts's/console.ts's own runQuery({rowsAsArray: true, ...}) call shape: every
// row scanned as []*string in column order, nil for SQL NULL, via B3's DatabaseTypeName-driven
// cellText — never a typeCast callback (the driver has none). The server-side kill is entirely
// adapter.go's Cancel (KILL QUERY over a side connection) — this function does not and must not
// try to make the query itself abort (B6, adapters.RunWithAbortRace's own contract).
func runArrayQuery(ctx context.Context, conn *sql.Conn, threadID uint32, query string, params []any, op *adapters.OpCtx, track TrackQuery, opts QueryOptions) ([][]*string, error) {
	setCommand(op, query, params, opts.LogParams)
	if err := adapters.CheckNotStarted(ctx); err != nil {
		return nil, err
	}

	release := track(RunningQuery{ThreadID: threadID})

	return adapters.RunWithAbortRace(ctx, release, func(queryCtx context.Context) ([][]*string, error) {
		rows, err := conn.QueryContext(queryCtx, query, params...)
		if err != nil {
			return nil, mapError(err)
		}
		defer rows.Close()

		types, err := rows.ColumnTypes()
		if err != nil {
			return nil, mapError(err)
		}
		dbTypes := make([]string, len(types))
		for i, t := range types {
			dbTypes[i] = t.DatabaseTypeName()
		}

		// raw/dest are pure scan scratch: rows.Scan fills raw in place, and cellText copies every
		// non-nil cell into a new Go string this same iteration, so neither needs to outlive it —
		// one allocation of each for the whole query instead of one pair per row (P2 R2, task #95).
		raw := make([]sql.RawBytes, len(types))
		dest := make([]any, len(types))
		for i := range raw {
			dest[i] = &raw[i]
		}

		var out [][]*string
		for rows.Next() {
			if err := rows.Scan(dest...); err != nil {
				return nil, mapError(err)
			}
			cells := make([]*string, len(types))
			for i, r := range raw {
				if r == nil {
					continue
				}
				text := cellText(r, dbTypes[i])
				cells[i] = &text
			}
			out = append(out, cells)
		}
		if err := rows.Err(); err != nil {
			return nil, mapError(err)
		}
		return out, nil
	})
}

// streamArrayQuery is runArrayQuery's single-pass twin (P2 R1): onRow is called once per scanned
// row, in order, with a []*string the callback owns, instead of every row being materialized into
// a [][]*string only to be transposed into the page builder immediately after. runArrayQuery
// itself stays as the array-returning shape countRows and the catalog paths want.
func streamArrayQuery(ctx context.Context, conn *sql.Conn, threadID uint32, query string, params []any, op *adapters.OpCtx, track TrackQuery, opts QueryOptions, onRow func(row []*string) error) error {
	setCommand(op, query, params, opts.LogParams)
	if err := adapters.CheckNotStarted(ctx); err != nil {
		return err
	}

	release := track(RunningQuery{ThreadID: threadID})

	_, err := adapters.RunWithAbortRace(ctx, release, func(queryCtx context.Context) (struct{}, error) {
		rows, err := conn.QueryContext(queryCtx, query, params...)
		if err != nil {
			return struct{}{}, mapError(err)
		}
		defer rows.Close()

		types, err := rows.ColumnTypes()
		if err != nil {
			return struct{}{}, mapError(err)
		}
		dbTypes := make([]string, len(types))
		for i, t := range types {
			dbTypes[i] = t.DatabaseTypeName()
		}

		// raw/dest are pure scan scratch — see runArrayQuery's identical comment above.
		raw := make([]sql.RawBytes, len(types))
		dest := make([]any, len(types))
		for i := range raw {
			dest[i] = &raw[i]
		}

		for rows.Next() {
			if err := rows.Scan(dest...); err != nil {
				return struct{}{}, mapError(err)
			}
			cells := make([]*string, len(types))
			for i, r := range raw {
				if r == nil {
					continue
				}
				text := cellText(r, dbTypes[i])
				cells[i] = &text
			}
			if err := onRow(cells); err != nil {
				return struct{}{}, err
			}
		}
		if err := rows.Err(); err != nil {
			return struct{}{}, mapError(err)
		}
		return struct{}{}, nil
	})
	return err
}

// CommandOptions is query.ts's CommandOptions.
type CommandOptions struct {
	// SuppressCommand: setCommand() was already called once for the whole batch (P5 D9) — do not
	// call it again.
	SuppressCommand bool
}

// runCommand is query.ts's runCommand: an UPDATE/DELETE/INSERT/START TRANSACTION/COMMIT/ROLLBACK
// has no rows worth returning, only the number of rows it affected.
func runCommand(ctx context.Context, conn *sql.Conn, threadID uint32, query string, params []any, op *adapters.OpCtx, track TrackQuery, opts CommandOptions) (int64, error) {
	if !opts.SuppressCommand {
		op.SetCommand(query)
	}
	if err := adapters.CheckNotStarted(ctx); err != nil {
		return 0, err
	}

	release := track(RunningQuery{ThreadID: threadID})

	return adapters.RunWithAbortRace(ctx, release, func(queryCtx context.Context) (int64, error) {
		result, err := conn.ExecContext(queryCtx, query, params...)
		if err != nil {
			return 0, mapError(err)
		}
		n, err := result.RowsAffected()
		if err != nil {
			return 0, mapError(err)
		}
		return n, nil
	})
}
